web/src/LogPaneLine.tsx

```diff
@@ -0,0 +1,159 @@
+import { Fields, BuildEvent } from "./types"
+import React, { PureComponent } from "react"
+import AnsiLine from "./AnsiLine"
+import {
+  Color,
+  ColorRGBA,
+  ColorAlpha,
+  SizeUnit,
+  Height,
+  Width,
+} from "./style-helpers"
+import styled from "styled-components"
+import Ansi from "ansi-to-react"
+
+type LogPaneProps = {
+  text: string
+  manifestName: string
+  level: string
+  fields?: Fields | null
+  lineId: number
+  shouldHighlight: boolean
+  showManifestPrefix: boolean
+  isContextChange: boolean
+}
+
+let LogPaneLineRoot = styled.span`
+  display: flex;
+
+  &.is-highlighted {
+    background-color: ${ColorRGBA(Color.blue, ColorAlpha.translucent)};
+  }
+
+  &.is-buildEvent-init {
+    margin-top: ${SizeUnit(0.5)};
+    margin-bottom: ${SizeUnit(0.5)};
+    background-color: ${Color.gray};
+    text-align: right;
+    padding-right: ${SizeUnit(1)};
+    border-top: 1px solid ${ColorRGBA(Color.grayLight, ColorAlpha.translucent)};
+    border-bottom: 1px solid
+      ${ColorRGBA(Color.grayLight, ColorAlpha.translucent)};
+  }
+  &.is-buildEvent-fallback {
+    background-color: ${Color.grayDarker};
+
+    // A lil' trick so bottom margin only appears after the last element
+    margin-top: -${SizeUnit(0.5)};
+    margin-bottom: ${SizeUnit(0.5)};
+  }
+`
+let LogLinePrefixRoot = styled.span`
+  user-select: none;
+  width: ${Width.secondaryNavItem}px; // Match height of tab above
+  box-sizing: border-box;
+  background-color: ${Color.grayDarker};
+  color: ${Color.grayLightest};
+  padding-left: ${SizeUnit(0.5)};
+  padding-right: ${SizeUnit(0.5)};
+  display: flex;
+  align-items: center;
+  justify-content: right;
+  flex-shrink: 0;
+  // Truncate long text:
+  text-overflow: ellipsis;
+  overflow: hidden;
+  white-space: nowrap;
+
+  ${LogPaneLineRoot}.is-contextchange > & {
+    margin-top: -${Height.logLineSeparator}px;
+    border-top: ${Height.logLineSeparator}px solid ${Color.gray};
+  }
+`
+
+let LineContent = styled(AnsiLine)`
```

Comment:
for what it's worth, in my performance investigation, I was finding that styled-components adds non-trivial overhead because it's basically wrapping every react component in an additional react component. 

It's not a big deal in most cases. But for LogLines (where there might be thousands or tens of thousands of them), I was seeing significant performance slow-down each time we added a new styled(). That's why I kept doing things with spans and CSS classes here.

Reply:
This is a very good point! I had wondered about this. I'll go ahead and use span and CSS here.